### ISSUE TITLE
[tf2.0.0] Fails to build on Python 3.8 - suggested fix (change nullptr to 0 in source)

### ISSUE THREAD

**Issue by dbonner** [tensorflow/tensorflow#33543]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): source
- TensorFlow version: 2.0.0 with cuda and TensorRT
- Python version: 3.8
- Installed using virtualenv? pip? conda?: virtualenv
- Bazel version (if compiling from source): 0.26.1
- GCC/Compiler version (if compiling from source): 7.4.0
- CUDA/cuDNN version: CUDA 10.0, CUDNN 7.6.4
- GPU model and memory: NVidia RTX 2080 Ti

**Describe the problem**

While building from source:
ERROR: /home/daniel/tensorflow/tensorflow/python/BUILD:341:1: C++ compilation of rule '//tensorflow/python:ndarray_tensor_bridge' failed (Exit 1) tensorflow/python/lib/core/ndarray_tensor_bridge.cc:108:1: error: cannot convert ‘std::nullptr_t’ to ‘Py_ssize_t {aka long int}’ in initialization

**Provide the exact sequence of commands / steps that you executed before running into the problem**

git checkout r2.0

bazel build --config=opt --config=cuda --config=v2 --cxxopt="-D_GLIBCXX_USE_CXX11_ABI=0" //tensorflow/tools/pip_package:build_pip_package

Fix:

In Python 3.8, the reserved "tp_print" slot was changed from a function pointer to a number, `Py_ssize_t tp_vectorcall_offset`.
In C, there is no "nullptr"; either a 0 or NULL casts automatically to both pointers and numbers.
Use 0 instead of "nullptr" in the slot to be source-compatible both with Python 3.8 and previous versions.

Search for "tp_print" in the files listed below.  Change nullptr to 0 where tp_print is in the comment:
i.e. 
change:
`    nullptr,                                      /* tp_print */     `
to:
`    0,                                      /* tp_print */     `
List of files that need this change:
/tensorflow/tensorflow/python/lib/core/ndarray_tensor_bridge.cc     ----> 1 change
/tensorflow/tensorflow/python/lib/core/bfloat16.cc                -----> 1 change
/tensorflow/tensorflow/python/eager/pywrap_tfe_src.cc      -------> 2 changes ( /* tp_print */ occurs twice)

**Comment by yongtang** [comment]
Thanks @dbonner. I tested on Ubuntu 18.04 + python 3.8 and the build completed successfully. Created a PR #33575 to carry the fix.

Note tensorflow depends on h5py which does not have python 3.8 support yet, as such I think a release version of tensorflow for python 3.8 may have to wait (for h5py).

**Comment by ahtik** [comment]
Also confirming that we managed to compile Tensorfow 2.0 from master (62a81c36ea, #33575 being merged) using Python 3.8 and CUDA 10.1 on Windows 10 without any tweaks/hacks. Couldn't test with CUDA 10.0 as it's been problematic with Win10). Build env must have numpy==1.17.3. 

XLA disabled, AFAIK on W10 it doesn't work out of the box/requires heavy tweaking.

**Comment by grisaitis** [comment]
@yongtang re h5py support for py 3.8, where are they tracking work for that? (I don't see an issue on their repo.) Also fwiw, i can pip install h5py on 3.8 just fine. Haven't tested it out, though.

**Comment by yongtang** [comment]
@grisaitis From the PyPI.org https://pypi.org/project/h5py/2.10.0/#files it looks like h5py does not have out-of-box binary release for python 3.8 on Linux, though the source is available, so install h5py with python 3.8 might work on some distributions.

**Comment by eidalex** [comment]
@ahtik Hi, would you please describe further in details the procedure you have taken to do so ? I have spent a whole day and a half trying to build it from source on : 
- Windows 10 Enterprise 1709
- Latest branch of Tensorflow 2.0 master
- CUDA 10.1 (cuDNN 7.6 & TensorRT 6)
- Python 3.8 with (abs-py 0.8.1, astor 0.8.0, h5py 2.10.0, keras-applications 1.0.8, kera-preprocessing 1.0.9, numpy 1.17.3, pip 19.3.1, setuptools 41.2.0)
- Visual Studio 2019 (tried MSCV 2019 as well as MSVC 2017)
- several versions of Bazel (0.26.0, 0.22.0, 0.29.1 even 1.1.0... well the latter did not work.. as expected though)
- MSYS2 20190524
- disabled XLA and ROCm

It is important to me as I would like to compile it to use AVX2, SSE4.x and FMA instructions set... Every time I run : `bazel build --define=no_tensorflow_py_deps=true //tensorflow/tools/pip_package:build_pip_package` it process, raise WARNINGS on imports in Bazel BUILD and after a certain amount of time I get:

> Target //tensorflow/tools/pip_package:build_pip_package failed to build
> INFO: Elapsed time: 2914.437s, Critical Path: 504.14s
> INFO: 2995 processes: 2995 local.
> FAILED: Build did NOT complete successfully

FYI : Bazel, MSYS, CUDA are perfectly working as standalones.
Thank you in advance for your time :)

**Comment by ahtik** [comment]
@eidalex Happy to.
First make sure to have bazel 0.26.1 and MSVC 2017 installed.
Then open the cmd.exe and set the env variables:
```bash
set BAZEL_VC=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC
set BAZEL_VS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community
set CUDA_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1
SET CUDA_TOOLKIT_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1
```
Check that your current python env is activated with python 3.8.0, my pip freeze outputs
`h5py==2.10.0
Keras-Applications==1.0.6
Keras-Preprocessing==1.0.5
numpy==1.17.3
six==1.12.0`.

I do have MSYS2 installed but think it's irrelevant for the build?

To build the package, I was using:
`bazel build //tensorflow/tools/pip_package:build_pip_package`

Please let me know if this fixes it, what might have been the issue.

**Comment by eidalex** [comment]
@ahtik You are my savior ! Thanks a lot !

Actually lets sumarize what happened there. Thank to you I have made the following modifications:
- Uninstall Visual Studio 2019 and install Visual Studio 2017 (the latter already contains all the right versions of the compiler MSVC)
- Delete the environment variable BAZEL_VC_FULL_VERSION which I had set (apparently it was optional so I don't think it made a real change)
- Adding CUDA_TOOLKIT_PATH variable 
- Installing the TensorFlow 2.0 with the special commit (62a81c3, #33575 being merged)

It builds successfully with AVX2 ! 

Note: I tried the latest version of the master and it did not compiled because of build configuration errors.(it also did not work with the r2.0 branch version)

> ERROR: error loading package '' : in C:/tensorflow/tensorflow/workspace.bzl : LABEL '//tensorflow/tools/def_file_filter_configure.bzl' crosses boundary of package 'tensorflow' (perhaps you want to put the column here: '//tensorflow:tools/def_file_filter/def_file_filter_configure.bzl')

Well.. I tried to modify the wrorkspace and the BUILD but it did not work at all...

**Comment by ahtik** [comment]
@eidalex If you want to build from the latest master (https://github.com/tensorflow/tensorflow/commit/eda53c63dab8b364872ede8e423e4fed5d1686f7), try reverting https://github.com/tensorflow/tensorflow/commit/3cdc97306c0a72a939b6444e00bc880f98a06a14

I don't know mlir enough to understand why it still asks for EnumsGen.h but reverting it makes the build succeed with both python 3.7 and 3.8. This build error happens only with Windows, not sure where or whether to report it or not... Strangely enough your error message looks a bit different than mine.

A quick way to test whether the workspace is affected by the mlir issue or not:
`bazel build //tensorflow/compiler/mlir/lite/quantization/lite:quantize_model`

**Comment by gunan** [comment]
I was able to successfully build TF on windows with python 3.8.
However, grpcio still did not publish their python 3.8 package.
Therefore, we are blocked on their release for our python 3.8 release

**Comment by IshanManchanda** [comment]
Grpcio has released their binaries for 3.8. Do we have an estimate for when TF will be releasing?

**Comment by eduard93** [comment]
[grpcio](https://pypi.org/project/grpcio/1.26.0/#files) and [h5py](https://pypi.org/project/h5py/2.10.0/#files) now provide binaries for python 3.8.

**Comment by gunan** [comment]
See the below post for the plans.
https://github.com/tensorflow/tensorflow/issues/33374#issuecomment-571074915

**Comment by hafez-ahmad** [comment]
is TensorFlow available for python 3.8? 
I was unable to install with python 3.8, do I need to downgrade into python 3.7?
